source/MRViewer/MRSaveOnClose.cpp

```diff
@@ -88,24 +89,27 @@ void SaveOnClosePlugin::preDraw_()
         {
             auto savePath = SceneRoot::getScenePath();
             if ( savePath.empty() )
-                savePath = saveFileDialog( { {}, {},SceneFileFilters } );
+                savePath = saveFileDialog( { {}, {}, SceneFileWriteFilters } );
 
-            ProgressBar::orderWithMainThreadPostProcessing( "Saving scene", [savePath, &root = SceneRoot::get(), viewer = Viewer::instance()]()->std::function<void()>
-            {
-                auto res = serializeObjectTree( root, savePath, ProgressBar::callBackSetProgress );
-                if ( !res.has_value() )
-                    spdlog::error( res.error() );
-
-                return[savePath, viewer, success = res.has_value()]()
+            ImGui::CloseCurrentPopup();
+            if ( !savePath.empty() )
+                ProgressBar::orderWithMainThreadPostProcessing( "Saving scene", [&shouldClose = shouldClose_, savePath, &root = SceneRoot::get(), viewer = Viewer::instance()]()->std::function<void()>
                 {
-                    if ( success )
+                    auto res = ObjectSave::toAnySupportedSceneFormat( root, savePath, ProgressBar::callBackSetProgress );
+
+                    return[&shouldClose = shouldClose, savePath, viewer, res]()
                     {
-                        viewer->onSceneSaved( savePath );
-                        glfwSetWindowShouldClose( Viewer::instance()->window, true );
-                    }
-                };
-            } );
-        }
+                        if ( res )
+                        {
+                            viewer->onSceneSaved( savePath );
+                            glfwSetWindowShouldClose( Viewer::instance()->window, true );
+                            shouldClose = true;
```

Comment:
did it work before without this line?

Reply:
No. The application would stay open after successful save.